src/Features/LanguageServer/Protocol/Handler/Diagnostics/DocumentPullDiagnosticHandler.cs

```diff
@@ -0,0 +1,154 @@
+            DocumentDiagnosticsParams request, RequestContext context, CancellationToken cancellationToken)
+        {
+            Contract.ThrowIfNull(request.TextDocument, $"Got a DocumentPullDiagnostic request that did not specify a {nameof(request.TextDocument)}");
+
+            var document = SolutionProvider.GetDocument(request.TextDocument, context.ClientName);
+            if (document == null)
+            {
+                // Client is asking server about a document that no longer exists (i.e. was removed/deleted from the
+                // workspace).  In that case we need to return an actual diagnostic report with `null` for the
+                // diagnostics to let the client know to dump that file entirely.
+                return new[] { new DiagnosticReport { ResultId = null, Diagnostics = null } };
+            }
+
+            var project = document.Project;
+            var solution = project.Solution;
+            var workspace = solution.Workspace;
+
+            // If the client has already asked for diagnostics for this document, see if we have actually recorded any
+            // differences, or if they should just use the same diagnostics as before.
+            if (request.PreviousResultId != null)
+            {
+                lock (_gate)
+                {
+                    if (_documentIdToLastResultId.TryGetValue((workspace, document.Id), out var lastReportedResultId) &&
+                        lastReportedResultId == request.PreviousResultId)
+                    {
+                        // Nothing changed between the last request and this one.  Report a null response to the client
+                        // to know they don't need to do anything.
+                        return null;
+                    }
+                }
+            }
+
+            // Being asked about this document for the first time.  Or being asked again and we have different diagnostics.
+
+            var diagnosticKeys = _diagnosticService.GetDiagnosticsUpdatedEventArgs(workspace, project.Id, document.Id, cancellationToken).ToImmutableArray();
+
+            using var _ = ArrayBuilder<LspDiagnostic>.GetInstance(out var result);
+
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            foreach (var key in diagnosticKeys)
+            {
+                var id = key.Id;
+                var documentId = key.DocumentId;
+
+                // ignore diagnostics unrelated to an actual document.
+                if (documentId == null)
+                    continue;
+
+                var diagnostics = _diagnosticService.GetDiagnostics(key.Workspace, key.ProjectId, key.DocumentId, key.Id, includeSuppressedDiagnostics: false, cancellationToken);
+                if (diagnostics == null)
+                    continue;
+
+                foreach (var diagnostic in diagnostics)
+                    result.AddIfNotNull(Convert(text, diagnostic));
```

Comment:
Shouldn't need AddIfNotNull, Convert method doesn't return null (see below annotation change)

Reply:
you are correct.  previously i returned null, but then switched to contracts.